packages/dev-server-hmr/src/HmrPlugin.ts

```diff
@@ -0,0 +1,184 @@
+import { Plugin, WebSocketsManager } from '@web/dev-server-core';
+import type { ServerArgs } from '@web/dev-server-core/dist/Plugin';
+import WebSocket from 'ws';
+import type { Context } from 'koa';
+
+export interface HmrReloadMessage {
+  type: 'reload';
+}
+
+export interface HmrUpdateMessage {
+  type: 'update';
+  url: string;
+}
+
+export type HmrMessage = HmrReloadMessage | HmrUpdateMessage;
+
+export interface HmrModule {
+  dependencies: Set<string>;
+  dependents: Set<string>;
+  hmrAccepted: boolean;
+  hmrEnabled: boolean;
+}
+
+export class HmrPlugin implements Plugin {
+  name = 'hmr';
+  injectWebSocket = true;
+
+  protected _dependencyTree: Map<string, HmrModule> = new Map();
+  protected _webSockets?: WebSocketsManager;
+
+  async serverStart({ webSockets, fileWatcher }: ServerArgs) {
+    if (!fileWatcher) {
+      throw new Error('Cannot use HMR when watch mode is disabled.');
+    }
+
+    if (!webSockets) {
+      throw new Error('Cannot use HMR when web sockets are disabled.');
+    }
+
+    this._webSockets = webSockets;
+
+    webSockets.on('message', ({ webSocket, data }) => this._onMessage(webSocket, data));
+    fileWatcher.on('change', path => this._onFileChanged(path));
+    fileWatcher.on('unlink', path => this._onFileChanged(path));
+  }
+
+  async transformImport({ source, code, context }: { source: string, code?: string, context: Context }) {
+    // If the module references import.meta.hot it can be assumed it
+    // supports hot reloading
+    const hmrEnabled = code?.includes('import.meta.hot') === true;
+    const mod = this._getOrCreateModule(context.path);
+    const dependencyMod = this._getOrCreateModule(source);
+    dependencyMod.hmrEnabled = hmrEnabled;
+
+    mod.dependencies.add(source);
+    dependencyMod.dependents.add(context.path);
+  }
+
+  async transform({ context }: { context: Context }) {
+    // TODO (43081j): won't work for now since transformImport runs first
```

Comment:
would it not be possible to store the dependency graph received from `transformImport`, and then afterwards in `transform` do the `hmrEnabled` check?

Reply:
if we can be sure thats the order it runs in, sure. though ill have to keep a reference around for the source of each module unless there's a way to get at it already 🤔

Reply:
Yep it's available as `context.body`